File: app/interfaces/web/frontend/src/api/roles.ts
import { z } from "zod"

export const AccountRoleSchema = z.union([
  z.literal("owner"),
  z.literal("admin"),
  z.literal("user"),
])
export const ManagedCreationRoleSchema = z.union([z.literal("admin"), z.literal("user")])

export type AccountRole = z.infer<typeof AccountRoleSchema>
export type ManagedCreationRole = z.infer<typeof ManagedCreationRoleSchema>

export const MAX_ADMINS = 5
export const MAX_ACCOUNTS = 16

const ROLE_RANK: Readonly<Record<AccountRole, number>> = {
  admin: 2,
  owner: 3,
  user: 1,
}

const ROLE_LIST_ORDER: Readonly<Record<AccountRole, number>> = {
  owner: 0,
  admin: 1,
  user: 2,
}

export function isManagerRole(role: AccountRole): role is "owner" | "admin" {
  return role === "owner" || role === "admin"
}

export function canManageRole(actorRole: AccountRole, targetRole: AccountRole): boolean {
  return ROLE_RANK[actorRole] > ROLE_RANK[targetRole]
}

export function compareAccountListOrder(
  left: { readonly role: AccountRole; readonly user_id: number },
  right: { readonly role: AccountRole; readonly user_id: number },
): number {
  return ROLE_LIST_ORDER[left.role] - ROLE_LIST_ORDER[right.role] || left.user_id - right.user_id
}

File: app/features/accounts/roles.py
"""Canonical account roles and strict hierarchy predicates."""

from __future__ import annotations

from typing import Final, Literal

from typing_extensions import TypeAlias

AccountRole: TypeAlias = Literal["owner", "admin", "user"]

MAX_ADMINS: Final[int] = 5
MAX_ACCOUNTS: Final[int] = 16
_MANAGER_ROLES: Final[frozenset[AccountRole]] = frozenset({"owner", "admin"})


class AccountRoleError(ValueError):
    """Raised when persisted role data is outside the canonical role set."""

    def __init__(self, raw_role: str) -> None:
        self.raw_role = raw_role
        super().__init__(raw_role)

    def __str__(self) -> str:
        return f"unsupported account role: {self.raw_role!r}"


def parse_account_role(raw_role: str) -> AccountRole:
    """Parse one role loaded from a persistence or HTTP boundary."""
    if raw_role == "owner":
        return "owner"
    if raw_role == "admin":
        return "admin"
    if raw_role == "user":
        return "user"
    raise AccountRoleError(raw_role)


def role_rank(role: AccountRole) -> int:
    """Return the immutable management rank for one canonical role."""
    if role == "owner":
        return 3
    if role == "admin":
        return 2
    if role == "user":
        return 1
    raise AccountRoleError(role)


def is_manager(role: AccountRole) -> bool:
    """Return whether a role may enter the shared management surface."""
    return role in _MANAGER_ROLES


def can_manage_role(actor_role: AccountRole, target_role: AccountRole) -> bool:
    """Return whether an actor may mutate a strictly lower role."""
    return role_rank(actor_role) > role_rank(target_role)


__all__ = (
    "AccountRole",
    "AccountRoleError",
    "MAX_ACCOUNTS",
    "MAX_ADMINS",
    "can_manage_role",
    "is_manager",
    "parse_account_role",
    "role_rank",
)
